### PR TITLE
[CSRanking] Adjust ranking of `makeIterator` overloads in `for-in` context

### LINKED ISSUE #60514: Sequence/Iterator conformance broken in 5.7
**Describe the bug**

Code utilizing `Sequence` and `Iterator` protocols that worked in previous versions of Swift no longer type-checks in 5.7 snapshots.

**Steps To Reproduce**

Try to build [WasmTransformer](https://github.com/swiftwasm/WasmTransformer) with Xcode 14.0 beta 5.

**Expected behavior**

This code type-checks and builds as it did Swift 5.5/5.6

**Screenshots**

Screenshot 2022-08-11 at 19 38 05

**Environment (please fill out the following information)**

 - OS: macOS 12.5 and Ubuntu 20.04
 - Xcode Version/Tag/Branch: 14.0 beta 5 (14A5294e), also reproducible on Linux with `swiftlang/swift:nightly-5.7-focal` image and `6fa8c6b08e85` hash.

**Additional context**

Here's the code snippet (not self-contained, but I hope it helps to pinpoint the issue) that causes these new errors:

```swift
public protocol VectorSectionReader: Sequence where Element == Result<Item, Error> {
    associatedtype Item
    var count: UInt32 { get }
    mutating func read() throws -> Item
}

public struct VectorSectionIterator<Reader: VectorSectionReader>: IteratorProtocol {
    private(set) var reader: Reader
    private(set) var left: UInt32
    init(reader: Reader, count: UInt32) {
        self.reader = reader
        self.left = count
    }
    private var end: Bool = false
    public mutating func next() -> Reader.Element? {
        guard !end else { return nil }
        guard left != 0 else { return nil }
        let result = Result(catching: { try reader.read() })
        left -= 1
        switch result {
        case .success: return result
        case .failure:
            end = true
            return result
        }
    }
}

extension VectorSectionReader {
    __consuming public func makeIterator() -> VectorSectionIterator<Self> {
        VectorSectionIterator(reader: self, count: count)
    }

    public func collect() throws -> [Item] {
        var items: [Item] = []
        items.reserveCapacity(Int(count))
        for result in self {
            try items.append(result.get())
        }
        return items
    }
}
```

### LINKED ISSUE #60958: [Regression] Ambiguous makeIterator() in 5.7
**Describe the bug**
The following snippet 
```swift
struct S: Sequence {
  private var _storage: [UInt8] = []
  func makeIterator() -> Array<UInt8>.Iterator {
    _storage.makeIterator()
  }
  
  typealias Element = UInt8
  class Iterator: IteratorProtocol {
    func next() -> UInt8? { 0 }
    typealias Element = UInt8
  }

  func makeIterator() -> Iterator {
    Iterator()
  }
}

extension S {
  func f() {
    for elt in self { //  error: ambiguous use of 'makeIterator()'
      print(elt)
    }
  }
}
```

**Steps To Reproduce**
1. Open this on Compiler Explorer https://godbolt.org/z/7seohEf8K
2. Note that in 5.6 that compiles but 5.7 and main nightly it gives as ambiguous `makeIterator()`
...

**Expected behavior**
Should compile

**Screenshots**

**Environment (please fill out the following information)**
 - OS: [e.g. macOS 11.0]
 - Xcode Version/Tag/Branch:

**Additional context**
```
...
Resolved overloads:
  selected overload set choice self: $T0 == S for locator@0xbbc9e48 [DeclRef@<source>:23:16]
  selected overload set choice $elt$generator: $T4 == @lvalue $T2 for locator@0xbbca298 [DeclRef@<invalid loc>]
  ($T6 [attributes: delayed] [involves_type_vars: $T8] [with possible bindings: (subtypes of) $T8?])
  (attempting disjunction choice $T1 bound to decl output.(file).S.makeIterator()@<source>:4:8 : (S) -> () -> Array<UInt8>.Iterator [[locator@0xbbc9eb8 [UnresolvedDot@<source>:23:16 -> member]]];
    (overload set choice binding $T1 := () -> Array<UInt8>.Iterator)
    (overload set choice binding $T5 := () -> $T9.Element?)
    (attempting disjunction choice UInt8? bind $T8? [deep equality] [[locator@0xbbca5c0 [DeclRef@<source>:23:16 -> sequence element type -> optional payload]]];
      ($T7 [with possible bindings: (supertypes of) UInt8])
      Initial bindings: $T7 := UInt8
      (attempting type variable $T7 := UInt8
        (found solution: <default 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>)
      )
    )
  )
  (attempting disjunction choice $T1 bound to decl output.(file).S.makeIterator()@<source>:14:8 : (S) -> () -> S.Iterator [[locator@0xbbc9eb8 [UnresolvedDot@<source>:23:16 -> member]]];
    (overload set choice binding $T1 := () -> S.Iterator)
    (overload set choice binding $T5 := () -> UInt8?)
    (attempting disjunction choice UInt8? bind $T8? [deep equality] [[locator@0xbbca5c0 [DeclRef@<source>:23:16 -> sequence element type -> optional payload]]];
      ($T7 [with possible bindings: (supertypes of) UInt8])
      Initial bindings: $T7 := UInt8
      (attempting type variable $T7 := UInt8
        (found solution: <default 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0>)
      )
    )
  )
  (attempting disjunction choice $T1 bound to decl Swift.(file).Sequence extension.makeIterator() : <Self where Self : Sequence, Self == Self.Iterator> (__owned Self) -> () -> Self [[locator@0xbbc9eb8 [UnresolvedDot@<source>:23:16 -> member]]];
    (failed constraint $T10 bind $T10.Iterator [[locator@0xbc10788 [UnresolvedDot@<source>:23:16 -> member -> opened generic -> type parameter requirement #1 (same-type)]]];)
    (overload set choice binding $T1 := () -> $T10)
  )
```
Maybe there is a bit more to it, but is a bit strange that the overload that is not the one from `Sequence` conformance is being considered here and even if considered how it is ending up with the same score.

### PR DESCRIPTION
**Pull request by xedin**
Prefer an overload of `makeIterator` that satisfies 
`Sequence#makeIterator` requirement  when `.makeIterator()`
is implicitly injected after sequence expression in `for-in` 
loop context.

This helps to avoid issues related to conditional conformances 
to `Sequence` that add `makeIterator` preferred by overloading 
rules that could change behavior i.e.:

```swift
extension Array where Element == Double {
  func makeIterator() -> Array<Int>.Iterator {
    return [1, 2, 3].makeIterator()
  }
}

for i in [4.0, 5.0, 6.0] {
  print(i)
}
```

`makeIterator` declared in the `Array` extension is not a witness to `Sequence#makeIterator` and shouldn't be used because that would change runtime behavior.

Resolves: https://github.com/apple/swift/issues/60514
Resolves: https://github.com/apple/swift/issues/60958

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by AnthonyLatsis — COMMENTED** [review]
Do you mind sticking to the `issue-*` format for file names? This is the one I'm planning to use once I start renaming `sr*` files.

**Comment by xedin** [comment]
@AnthonyLatsis Sure, let's let the source compat run and I'm going to rename before merging.

File: validation-test/Sema/issue60958.swift [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// RUN: %target-swift-frontend -emit-silgen %s | %FileCheck %s
+
+// https://github.com/apple/swift/issues/60958
+// Make sure that `makeIterator` witness is picked over the non-witness.
+
+struct S: Sequence {
```
Comment (LucianoPAlmeida) [inline validation-test/Sema/issue60958.swift:6]: @xedin Could be worth adding a test case that involves a generic argument?
```swift
struct G<T>: Sequence {
  private var _storage: [T] = []
  func makeIterator() -> Array<T>.Iterator {
    _storage.makeIterator()
  }
  
  typealias Element = T
  class Iterator: IteratorProtocol {
    func next() -> T? { fatalError() }
    typealias Element = T
  }

  func makeIterator() -> Iterator {
    Iterator()
  }
}

extension G {
  func f() {
    for elt in self { // OK
      print(elt)
    }
  }
}
```

Comment (xedin) [reply validation-test/Sema/issue60958.swift:6]: Other test-case is generic actually :)

Comment (LucianoPAlmeida) [reply validation-test/Sema/issue60958.swift:6]: Ah true, miss that 😅

**Pull request closed without merging**